akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala

```diff
@@ -292,13 +288,19 @@ private[projection] abstract class InternalProjectionState[Offset, Envelope](
               handler.process(context.envelope).map { done =>
                 // `telemetry.afterProcess` is invoked immediately after `handler.process`
                 telemetry.afterProcess(context.externalContext)
+
+                try {
+                  statusObserver.offsetProgress(projectionId, context.envelope)
+                } catch {
+                  case NonFatal(_) => // ignore
+                }
                 telemetry.onOffsetStored(1)
+
                 done
               }
             }
-            val processed =
-              handlerRecovery.applyRecovery(context.envelope, context.offset, context.offset, abort.future, measured)
-            reportProgress(processed, context.envelope)
```

Comment:
This is the trickiest case in the whole PR and it actually required writing a new test and provided a fix.

When using `exactly-once` and a retry strategy that may `skip` (e.g. `retryAndSkip` or plain `skip`)  the previous implementation might report progress wrongly:

 - the envelope `4` comes in, the handling fails
 - the maximum retries are reached and the envelope `4` is skipped
 - `statusObserver.reportProgress(4)` is invoked even though the transaction in the `handler` didn't succeed and the offset `4`was not stored (it was skipped).

The problem, the, was if the projection crashed immediately after reporting the offset `4`. The restarted projection would try and process the envelope `4` again (since the offset was not stored) and immediately after `statusObserver.reportProgress(4)` would be invoked again.